src/parsing.js

```diff
@@ -14,14 +14,78 @@
 
 (function(internal, testing) {
 
-  var _numberValueRegex = /^\s*[-+]?(\d*\.)?\d+(e[-+]?\d+)?\s*$/;
+  // Extra backslashes because otherwise JS interprets them incorrectly.
+  var _numberValueRegexStr = '[-+]?(\\d*\\.)?\\d+(e[-+]?\\d+)?';
 
   function isNumberValueString(cssString) {
-    return _numberValueRegex.test(cssString);
+    // Anchor the regex to the start and end of the string..
+    var numberValueRegex = new RegExp('^\\s*' + _numberValueRegexStr + '\\s*$');
+    return numberValueRegex.test(cssString);
+  }
+
+  function isCalc(string) {
+    return /^calc/.test(string); 
+  }
+
+  // Effectively returns a calc dictionary.
+  function parseDimension(unitRegExp, string) {
+    string = string.trim().toLowerCase();
+
+    // CSS allows lengths to be '0', so if px is a supported unit, return 0px.
+    if (string == '0' && 'px'.search(unitRegExp) >= 0)
+      return {px: 0};
+
+    // If we have parenthesis, we're a calc and need to start with 'calc'.
+    if (!/^[^(]*$|^calc/.test(string)) {
+      return;
+    }
+    string = string.replace(/^calc\(/, '(');
+
+    // We tag units by prefixing them with 'U' (note that we are already
+    // lowercase) to prevent problems with types which are substrings of
+    // each other (although prefixes may be problematic!)
+    var matchedUnits = {};
+    string = string.replace(unitRegExp, function(match) {
+      matchedUnits[match] = null;
+      return 'U' + match;
+    });
+    var taggedUnitRegExp = 'U(' + unitRegExp.source + ')';
+
+    // Validating input is simply applying as many reductions as we can.
+    var typeCheck = string.replace(new RegExp(_numberValueRegexStr, 'g'), 'N')
+                          .replace(new RegExp('N' + taggedUnitRegExp, 'g'), 'D')
+                          .replace(/\s[+-]\s/g, 'O')
+                          .replace(/\s/g, '');
+    var reductions = [/N\*(D)/g, /(N|D)[*/]N/g, /(N|D)O\1/g, /\((N|D)\)/g];
+    var i = 0;
+    while (i < reductions.length) {
+      if (reductions[i].test(typeCheck)) {
+        typeCheck = typeCheck.replace(reductions[i], '$1');
+        i = 0;
+      } else {
+        i++;
+      }
+    }
+    if (typeCheck != 'D') {
+      return;
+    }
+
+    for (var unit in matchedUnits) {
+      // TODO: Don't use eval here as it will throw up security flags.
```

Comment:
It's non-trivial because this method takes advantage of the behaviour of eval. We'd need to write a cumulating parser for them...